packages/backend/src/services/decision-loop/decision-loop-service.ts

```diff
@@ -212,33 +205,52 @@ export async function* runDecisionLoop(
   }
 }
 
-function removeTamboToolParameters(
+/**
+ * Build a tool call request from a tool call, even if it's incomplete.
+ * Returns undefined if the tool call is not a function type.
+ * Uses partial-json parsing to handle incomplete JSON during streaming.
+ */
+function buildToolCallRequest(
   toolCall: OpenAI.Chat.Completions.ChatCompletionMessageToolCall,
   tools: OpenAI.Chat.Completions.ChatCompletionTool[],
-  chunk: Partial<LLMResponse>,
-) {
-  const originalRequest = getLLMResponseToolCallRequest(chunk);
-  // Just means the tool call is still streaming
-  if (!originalRequest) {
-    return;
-  }
+): ToolCallRequest | undefined {
   // "custom" tool calls are not supported
   if (toolCall.type !== "function") {
-    return originalRequest;
+    return undefined;
   }
-  const parsedToolCall = tryParseJsonObject(toolCall.function.arguments, false);
-  if (parsedToolCall) {
-    const filteredArgs = filterOutStandardToolParameters(
-      toolCall,
-      tools,
-      parsedToolCall,
-    );
-
-    if (filteredArgs) {
-      return {
-        ...originalRequest,
-        parameters: filteredArgs,
-      };
+
+  // Try to parse the arguments (may be incomplete during streaming)
+  let parsedArgs: Record<string, unknown> | undefined;
+  try {
+    // Use partial-json parse to handle incomplete JSON during streaming
+    const parsed = parse(toolCall.function.arguments);
+    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
+      return undefined;
     }
+
+    parsedArgs = parsed;
+  } catch (_e) {
+    // If parsing fails completely, we can't build a request
+    return undefined;
   }
+
+  if (!parsedArgs) {
```

Comment:
all of this feels kinda redundant with the `if (!parsed || ...)` above - perhaps just move that down here?